spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMerge.java

```diff
@@ -0,0 +1,336 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.spark.extensions;
+
+import java.util.Arrays;
+import java.util.Map;
+import org.apache.iceberg.AssertHelpers;
+import org.apache.spark.sql.AnalysisException;
+import org.junit.After;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public abstract class TestMerge extends SparkRowLevelOperationsTestBase {
+
+  public TestMerge(String catalogName, String implementation, Map<String, String> config,
+                   String fileFormat, boolean vectorized) {
+    super(catalogName, implementation, config, fileFormat, vectorized);
+  }
+
+  @BeforeClass
+  public static void setupSparkConf() {
+    spark.conf().set("spark.sql.shuffle.partitions", "4");
+  }
+
+  @After
+  public void removeTables() {
+    sql("DROP TABLE IF EXISTS %s", tableName);
+    sql("DROP TABLE IF EXISTS source");
+  }
+
+  // TODO: tests for reordering when operations succeed (both insert and update actions)
+  // TODO: tests for modifying fields in a null struct
```

Comment:
What is the correct behavior here? Reject it?

I think what would currently happen is it would fail if any existing field were accessed, but it would succeed if all fields were assigned. Is that correct?

Reply:
I think rejecting will be too restrictive. That would mean we won't be able to update any nullable struct.

Here is what Postgres does:

```
postgres=# INSERT INTO t_100 (id, c) VALUES (2, null);
INSERT 0 1

postgres=# SELECT * FROM t_100;
 id |   c   
----+-------
  1 | (1,2)
  2 | 
(2 rows)

postgres=# UPDATE t_100 SET c.n1 = -1;
UPDATE 2

postgres=# SELECT * FROM t_100;
 id |   c    
----+--------
  1 | (-1,2)
  2 | (-1,)
(2 rows)

```

Reply:
I think the existing logic matches Postgres.

Reply:
Okay, so the struct gets created and unspecified fields get set to null. Sounds good to me.

Reply:
Yeah, correct.